p2p/peer_set_test.go

```diff
@@ -65,3 +70,49 @@ func TestAddRemoveMany(t *testing.T) {
 		}
 	}
 }
+
+func TestPeerSetAddDuplicate(t *testing.T) {
+	t.Parallel()
+	peerSet := NewPeerSet()
+	peer := randPeer()
+	if err := peerSet.Add(peer); err != nil {
+		t.Fatalf("Failed to add new peer: %v", err)
+	}
+
+	var wg sync.WaitGroup
+	for i := 0; i < 10; i++ {
+		// Add them asynchronously to test the
+		// concurrent guarantees of our APIs.
```

Comment:
Is this really testing anything useful in the "AddDuplicate" case ?

Reply:
What I had in mind was ensuring that we having at any point multiple goroutines all trying to add duplicates: just 4 goroutines going on might not cut it, so I chose 10.

Reply:
Maybe it'd be more effective if we ran all the Add in a go-routine, and counted the results, and assert they're all the DuplicatePeer error except one.

Reply:
So currently, that's what the test does. Counting the tests in separate goroutines requires synchronization for the counting which might make them unnecessarily complex(e.g a channel collecting results, then later counting each) IMO, and at the beginning I test that no error occurs the first time. However, if that eases reading I'll update it and after review we can decide which one we wanna pick :)

Reply:
That was actually a great suggestion that I realized while writing the test, thank you @ebuchman!

Reply:
By the above, I mean I updated the AddDuplicate case.